src/common/bpf.rs

```diff
@@ -10,6 +10,71 @@ pub struct BPF {
 #[cfg(not(feature = "bpf"))]
 pub struct BPF {}
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub enum ProbeType {
+    Kernel,
+    User,
+    Tracepoint,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub enum ProbeLocation {
+    Entry,
+    Return,
+}
+
+// Define a probe.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg(feature = "bpf")]
+pub struct FunctionProbe {
+    pub name: String,                  // name of the function to probe
+    pub handler: String,               // name of the probe handler
+    pub probe_type: ProbeType,         // function type, kernel, user or tracepoint.
+    pub probe_location: ProbeLocation, // probe location, at entry or at return.
+    pub binary_path: Option<String>,   // required for user probe only.
+    pub sub_system: Option<String>,    // required for tracepoint only.
+}
+
+#[cfg(feature = "bpf")]
+impl FunctionProbe {
+    // try to attach to a bpf instance.
+    #[allow(unused_assignments)]
+    pub fn try_attach_to_bpf(&self, bpf: &mut bcc::BPF) -> Result<(), anyhow::Error> {
+        match self.probe_type {
+            ProbeType::Kernel => match self.probe_location {
+                ProbeLocation::Entry => bcc::Kprobe::new()
+                    .handler(self.handler.as_str())
+                    .function(self.name.as_str())
+                    .attach(bpf)?,
+                ProbeLocation::Return => bcc::Kretprobe::new()
+                    .handler(self.handler.as_str())
+                    .function(self.name.as_str())
+                    .attach(bpf)?,
+            },
+            ProbeType::User => match self.probe_location {
+                ProbeLocation::Entry => bcc::Uprobe::new()
+                    .handler(self.handler.as_str())
+                    .binary(self.binary_path.as_ref().unwrap().as_str())
```

Comment:
This unwrap will mean we terminate the runtime if the binary path is not defined. While this does indicate there's an issue in the code, I'd rather at least use expect with some message about why, or better yet, log an error and skip this probe. If we use the error + skip approach, it should use the 'fault_tolerant' pattern used elsewhere so that we can make it a hard error in CI.

Reply:
I made an assumption that when an implementor tried to create a User probe, he will specify a binary path, so I did not handle the None case for the binrary_path here. 
I've made a change to log an error here.

Reply:
Handled the sub system for tracepoint too